`keyof this` is computed too narrowly on a subclass that passes itself as its own type parameter, causing TS2345 / TS2322 variance failure

## Steps to reproduce

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "strict": true,
    "target": "ES2022",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "noEmit": true
  }
}
```

`index.ts`:

```ts
declare class STModel<TAttrs extends {} = any, TCreate extends {} = TAttrs> {
  $add: (propertyKey: string) => void;
  $set: (propertyKey: keyof this) => void;
  data: TAttrs;
  newAttrs: TCreate;
}

interface IBase { id: string }

type Attrs<T extends {}> = { [K in keyof T]: T[K] } & IBase;
type CreateAttrs<T extends {}> = Partial<Attrs<T>>;

abstract class BaseModel<MA extends {}> extends STModel<Attrs<MA>, CreateAttrs<MA>> {}

class _Foo extends BaseModel<_Foo> {
  declare name: string;
}

const ctor: new () => STModel = _Foo;
console.log(ctor);
```

Tested with `@typescript/native-preview@beta` (`Version 7.0.0-dev.20260421.2`) and `typescript@6.0.3`.

This shape comes up in real code with `sequelize-typescript`'s `Model<TAttrs, TCreate>` plus an in-house `BaseModel<MA>` wrapper that maps the model's attributes through `Attrs<MA>` / `Partial<Attrs<MA>>` and is used as `class _Foo extends BaseModel<_Foo>`.

## Behavior with `typescript@6.0`

Compiles cleanly (exit 0, no diagnostics).

## Behavior with `tsgo`

```
index.ts(20,7): error TS2322: Type 'typeof _Foo' is not assignable to type 'new () => STModel<any, any>'.
  Type '_Foo' is not assignable to type 'STModel<any, any>'.
    Types of property '$set' are incompatible.
      Type '(propertyKey: "name") => void' is not assignable to type '(propertyKey: keyof STModel<any, any>) => void'.
        Types of parameters 'propertyKey' and 'propertyKey' are incompatible.
          Type 'keyof STModel<any, any>' is not assignable to type '"name"'.
            Type '"$add"' is not assignable to type '"name"'.
```

`keyof this` for `_Foo`'s `$set` is computed as just `"name"` — the *additionally declared* property — instead of `"$add" | "$set" | "data" | "newAttrs" | "name"`. The variance check on `(propertyKey: keyof this) => void` then fails because `"$add"` (from the parent's `keyof STModel<any, any>`) isn't assignable to `"name"`.

Removing any of the following pieces makes the error go away:

- The `BaseModel` intermediate (`class _Foo extends STModel<…>` directly).
- The `Partial` wrapper on `CreateAttrs`.
- The intersection with `IBase` inside `Attrs<T>`.
- Passing `_Foo` as its own type parameter (e.g. `class _Foo extends BaseModel<{ name: string }>`).

So the trigger is the combination of: a self-referential subclass type parameter, an intermediate generic class, and a mapped-then-intersected-then-`Partial`'d type used as the second type argument. tsc handles this combination correctly; tsgo apparently treats `keyof this` for the subclass instance type as only the locally declared properties (`"name"`) instead of including inherited members.

A very real-world impact: this means `class _Foo extends BaseModel<_Foo>` cannot be passed to a `(model: new () => Model) => void` registration function under tsgo, even though it can under tsc.

--- a/_packages/api/src/enums/objectFlags.enum.ts
+++ b/_packages/api/src/enums/objectFlags.enum.ts
@@ -36,6 +36,7 @@ export enum ObjectFlags {
     IsClassInstanceClone = 1 << 26,
     IdenticalBaseTypeCalculated = 1 << 27,
     IdenticalBaseTypeExists = 1 << 28,
+    UnresolvedMembers = 1 << 29,
     IsGenericTypeComputed = 1 << 22,
     IsGenericObjectType = 1 << 23,
     IsGenericIndexType = 1 << 24,

--- a/_packages/api/src/enums/objectFlags.ts
+++ b/_packages/api/src/enums/objectFlags.ts
@@ -36,6 +36,7 @@ export var ObjectFlags: any;
     ObjectFlags[ObjectFlags["IsClassInstanceClone"] = 67108864] = "IsClassInstanceClone";
     ObjectFlags[ObjectFlags["IdenticalBaseTypeCalculated"] = 134217728] = "IdenticalBaseTypeCalculated";
     ObjectFlags[ObjectFlags["IdenticalBaseTypeExists"] = 268435456] = "IdenticalBaseTypeExists";
+    ObjectFlags[ObjectFlags["UnresolvedMembers"] = 536870912] = "UnresolvedMembers";
     ObjectFlags[ObjectFlags["IsGenericTypeComputed"] = 4194304] = "IsGenericTypeComputed";
     ObjectFlags[ObjectFlags["IsGenericObjectType"] = 8388608] = "IsGenericObjectType";
     ObjectFlags[ObjectFlags["IsGenericIndexType"] = 16777216] = "IsGenericIndexType";

--- a/internal/checker/checker.go
+++ b/internal/checker/checker.go
@@ -233,9 +233,10 @@ type IterationTypesKey struct {
 // PropertiesTypesKey
 
 type PropertiesTypesKey struct {
-	typeId        TypeId
-	include       TypeFlags
-	includeOrigin bool
+	typeId            TypeId
+	include           TypeFlags
+	includeOrigin     bool
+	unresolvedMembers bool
 }
 
 // NonExistentPropertyKey
@@ -18654,6 +18655,7 @@ func (c *Checker) resolveObjectTypeMembers(t *Type, source *Type, typeParameters
 		}
 		c.setStructuredTypeMembers(t, members, callSignatures, constructSignatures, indexInfos)
 		thisArgument := core.LastOrNil(typeArguments)
+		t.objectFlags |= ObjectFlagsUnresolvedMembers
 		for _, baseType := range baseTypes {
 			instantiatedBaseType := baseType
 			if thisArgument != nil {
@@ -18672,6 +18674,7 @@ func (c *Checker) resolveObjectTypeMembers(t *Type, source *Type, typeParameters
 				return findIndexInfo(indexInfos, info.keyType) == nil
 			}))
 		}
+		t.objectFlags &^= ObjectFlagsUnresolvedMembers
 	}
 	c.setStructuredTypeMembers(t, members, callSignatures, constructSignatures, indexInfos)
 }
@@ -26163,7 +26166,7 @@ func (c *Checker) getExtractStringType(t *Type) *Type {
 }
 
 func (c *Checker) getLiteralTypeFromProperties(t *Type, include TypeFlags, includeOrigin bool) *Type {
-	key := PropertiesTypesKey{typeId: t.id, include: include, includeOrigin: includeOrigin}
+	key := PropertiesTypesKey{typeId: t.id, include: include, includeOrigin: includeOrigin, unresolvedMembers: t.objectFlags&ObjectFlagsUnresolvedMembers != 0}
 	if cached, ok := c.propertiesTypes[key]; ok {
 		return cached
 	}

--- a/internal/checker/types.go
+++ b/internal/checker/types.go
@@ -497,6 +497,7 @@ const (
 	// Flags that require TypeFlags.Object and ObjectFlags.Reference
 	ObjectFlagsIdenticalBaseTypeCalculated = 1 << 27 // has had `getSingleBaseForNonAugmentingSubtype` invoked on it already
 	ObjectFlagsIdenticalBaseTypeExists     = 1 << 28 // has a defined cachedEquivalentBaseType member
+	ObjectFlagsUnresolvedMembers           = 1 << 29 // Member resolution in process
 	// Flags that require TypeFlags.UnionOrIntersection or TypeFlags.Substitution
 	ObjectFlagsIsGenericTypeComputed = 1 << 22 // IsGenericObjectType flag has been computed
 	ObjectFlagsIsGenericObjectType   = 1 << 23 // Union or intersection contains generic object type

--- a/testdata/baselines/reference/compiler/keyofUnresolvedBaseMembers.symbols
+++ b/testdata/baselines/reference/compiler/keyofUnresolvedBaseMembers.symbols
@@ -1,0 +1,77 @@
+//// [tests/cases/compiler/keyofUnresolvedBaseMembers.ts] ////
+
+=== keyofUnresolvedBaseMembers.ts ===
+// https://github.com/microsoft/typescript-go/issues/3494
+
+declare class STModel<TAttrs extends {} = any, TCreate extends {} = TAttrs> {
+>STModel : Symbol(STModel, Decl(keyofUnresolvedBaseMembers.ts, 0, 0))
+>TAttrs : Symbol(TAttrs, Decl(keyofUnresolvedBaseMembers.ts, 2, 22))
+>TCreate : Symbol(TCreate, Decl(keyofUnresolvedBaseMembers.ts, 2, 46))
+>TAttrs : Symbol(TAttrs, Decl(keyofUnresolvedBaseMembers.ts, 2, 22))
+
+  $add: (propertyKey: string) => void;
+>$add : Symbol(STModel.$add, Decl(keyofUnresolvedBaseMembers.ts, 2, 77))
+>propertyKey : Symbol(propertyKey, Decl(keyofUnresolvedBaseMembers.ts, 3, 9))
+
+  $set: (propertyKey: keyof this) => void;
+>$set : Symbol(STModel.$set, Decl(keyofUnresolvedBaseMembers.ts, 3, 38))
+>propertyKey : Symbol(propertyKey, Decl(keyofUnresolvedBaseMembers.ts, 4, 9))
+
+  data: TAttrs;
+>data : Symbol(STModel.data, Decl(keyofUnresolvedBaseMembers.ts, 4, 42))
+>TAttrs : Symbol(TAttrs, Decl(keyofUnresolvedBaseMembers.ts, 2, 22))
+
+  newAttrs: TCreate;
+>newAttrs : Symbol(STModel.newAttrs, Decl(keyofUnresolvedBaseMembers.ts, 5, 15))
+>TCreate : Symbol(TCreate, Decl(keyofUnresolvedBaseMembers.ts, 2, 46))
+}
+
+interface IBase { id: string }
+>IBase : Symbol(IBase, Decl(keyofUnresolvedBaseMembers.ts, 7, 1))
+>id : Symbol(IBase.id, Decl(keyofUnresolvedBaseMembers.ts, 9, 17))
+
+type Attrs<T extends {}> = { [K in keyof T]: T[K] } & IBase;
+>Attrs : Symbol(Attrs, Decl(keyofUnresolvedBaseMembers.ts, 9, 30))
+>T : Symbol(T, Decl(keyofUnresolvedBaseMembers.ts, 11, 11))
+>K : Symbol(K, Decl(keyofUnresolvedBaseMembers.ts, 11, 30))
+>T : Symbol(T, Decl(keyofUnresolvedBaseMembers.ts, 11, 11))
+>T : Symbol(T, Decl(keyofUnresolvedBaseMembers.ts, 11, 11))
+>K : Symbol(K, Decl(keyofUnresolvedBaseMembers.ts, 11, 30))
+>IBase : Symbol(IBase, Decl(keyofUnresolvedBaseMembers.ts, 7, 1))
+
+type CreateAttrs<T extends {}> = Partial<Attrs<T>>;
+>CreateAttrs : Symbol(CreateAttrs, Decl(keyofUnresolvedBaseMembers.ts, 11, 60))
+>T : Symbol(T, Decl(keyofUnresolvedBaseMembers.ts, 12, 17))
+>Partial : Symbol(Partial, Decl(lib.es5.d.ts, --, --))
+>Attrs : Symbol(Attrs, Decl(keyofUnresolvedBaseMembers.ts, 9, 30))
+>T : Symbol(T, Decl(keyofUnresolvedBaseMembers.ts, 12, 17))
+
+abstract class BaseModel<MA extends {}> extends STModel<Attrs<MA>, CreateAttrs<MA>> {}
+>BaseModel : Symbol(BaseModel, Decl(keyofUnresolvedBaseMembers.ts, 12, 51))
+>MA : Symbol(MA, Decl(keyofUnresolvedBaseMembers.ts, 14, 25))
+>STModel : Symbol(STModel, Decl(keyofUnresolvedBaseMembers.ts, 0, 0))
+>Attrs : Symbol(Attrs, Decl(keyofUnresolvedBaseMembers.ts, 9, 30))
+>MA : Symbol(MA, Decl(keyofUnresolvedBaseMembers.ts, 14, 25))
+>CreateAttrs : Symbol(CreateAttrs, Decl(keyofUnresolvedBaseMembers.ts, 11, 60))
+>MA : Symbol(MA, Decl(keyofUnresolvedBaseMembers.ts, 14, 25))
+
+class _Foo extends BaseModel<_Foo> {
+>_Foo : Symbol(_Foo, Decl(keyofUnresolvedBaseMembers.ts, 14, 86))
+>BaseModel : Symbol(BaseModel, Decl(keyofUnresolvedBaseMembers.ts, 12, 51))
+>_Foo : Symbol(_Foo, Decl(keyofUnresolvedBaseMembers.ts, 14, 86))
+
+  declare name: string;
+>name : Symbol(_Foo.name, Decl(keyofUnresolvedBaseMembers.ts, 16, 36))
+}
+
+const ctor: new () => STModel = _Foo;
+>ctor : Symbol(ctor, Decl(keyofUnresolvedBaseMembers.ts, 20, 5))
+>STModel : Symbol(STModel, Decl(keyofUnresolvedBaseMembers.ts, 0, 0))
+>_Foo : Symbol(_Foo, Decl(keyofUnresolvedBaseMembers.ts, 14, 86))
+
+console.log(ctor);
+>console.log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>console : Symbol(console, Decl(lib.dom.d.ts, --, --))
+>log : Symbol(Console.log, Decl(lib.dom.d.ts, --, --))
+>ctor : Symbol(ctor, Decl(keyofUnresolvedBaseMembers.ts, 20, 5))
+

--- a/testdata/baselines/reference/compiler/keyofUnresolvedBaseMembers.types
+++ b/testdata/baselines/reference/compiler/keyofUnresolvedBaseMembers.types
@@ -1,0 +1,55 @@
+//// [tests/cases/compiler/keyofUnresolvedBaseMembers.ts] ////
+
+=== keyofUnresolvedBaseMembers.ts ===
+// https://github.com/microsoft/typescript-go/issues/3494
+
+declare class STModel<TAttrs extends {} = any, TCreate extends {} = TAttrs> {
+>STModel : STModel<TAttrs, TCreate>
+
+  $add: (propertyKey: string) => void;
+>$add : (propertyKey: string) => void
+>propertyKey : string
+
+  $set: (propertyKey: keyof this) => void;
+>$set : (propertyKey: keyof this) => void
+>propertyKey : keyof this
+
+  data: TAttrs;
+>data : TAttrs
+
+  newAttrs: TCreate;
+>newAttrs : TCreate
+}
+
+interface IBase { id: string }
+>id : string
+
+type Attrs<T extends {}> = { [K in keyof T]: T[K] } & IBase;
+>Attrs : Attrs<T>
+
+type CreateAttrs<T extends {}> = Partial<Attrs<T>>;
+>CreateAttrs : Partial<Attrs<T>>
+
+abstract class BaseModel<MA extends {}> extends STModel<Attrs<MA>, CreateAttrs<MA>> {}
+>BaseModel : BaseModel<MA>
+>STModel : STModel<Attrs<MA>, Partial<Attrs<MA>>>
+
+class _Foo extends BaseModel<_Foo> {
+>_Foo : _Foo
+>BaseModel : BaseModel<_Foo>
+
+  declare name: string;
+>name : string
+}
+
+const ctor: new () => STModel = _Foo;
+>ctor : new () => STModel
+>_Foo : typeof _Foo
+
+console.log(ctor);
+>console.log(ctor) : void
+>console.log : (...data: any[]) => void
+>console : Console
+>log : (...data: any[]) => void
+>ctor : new () => STModel
+

--- a/testdata/tests/cases/compiler/keyofUnresolvedBaseMembers.ts
+++ b/testdata/tests/cases/compiler/keyofUnresolvedBaseMembers.ts
@@ -1,0 +1,24 @@
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/3494
+
+declare class STModel<TAttrs extends {} = any, TCreate extends {} = TAttrs> {
+  $add: (propertyKey: string) => void;
+  $set: (propertyKey: keyof this) => void;
+  data: TAttrs;
+  newAttrs: TCreate;
+}
+
+interface IBase { id: string }
+
+type Attrs<T extends {}> = { [K in keyof T]: T[K] } & IBase;
+type CreateAttrs<T extends {}> = Partial<Attrs<T>>;
+
+abstract class BaseModel<MA extends {}> extends STModel<Attrs<MA>, CreateAttrs<MA>> {}
+
+class _Foo extends BaseModel<_Foo> {
+  declare name: string;
+}
+
+const ctor: new () => STModel = _Foo;
+console.log(ctor);

